linode/manager
# upcoming: [M3-7777, M3-7778] - Improve Billing & Account Restricted User Experience

## Description 📝
For the three account options: "Make a Payment," "Update Billing Contact Information," and "Add Payment Method," access will be disabled for restricted users or child accounts. Furthermore, parent account users will not have the ability to update the company name.

>[!note]
We have separate tickets for additional E2E test coverage [M3-7797] 👍. 

## Changes  🔄
List any change relevant to the reviewer.
- Removed `getDisabledTooltipText` util and replaced it with our existing `useRestrictedGlobalGrantCheck` and `getRestrictedResourceText` util.

## Target release date 🗓️
3/4/2024

## Preview 📷
| Before  | After: User Type = Default   | After: User Type = Child | After: User Type = Parent |
| ------- | ------- |------- |------- |
| ![Screenshot 2024-02-16 at 12 00 46 PM](https://github.com/linode/manager/assets/125309814/13f1b7aa-a139-4f1d-a193-f805d70ec7e4) | ![Screenshot 2024-02-16 at 12 00 24 PM](https://github.com/linode/manager/assets/125309814/65909b3c-6bb1-4a2f-a39a-0c61438f3192) |![Screenshot 2024-02-16 at 2 37 22 PM](https://github.com/linode/manager/assets/125309814/8247acff-dabd-4a4f-ac4c-3f4d9d7154b9) | ![Screenshot 2024-02-20 at 5 30 55 PM](https://github.com/linode/manager/assets/125309814/18ebe507-318a-4f35-a3e3-d97816ff0957) |
| ![Screenshot 2024-02-16 at 12 26 14 PM](https://github.com/linode/manager/assets/125309814/9bba044e-33b4-4334-b7f7-c5a3a37e6ebf) | ![Screenshot 2024-02-16 at 12 48 53 PM](https://github.com/linode/manager/assets/125309814/97c184f9-94c8-45ec-9e9d-cb2e1e66df94) |![Screenshot 2024-02-16 at 2 37 40 PM](https://github.com/linode/manager/assets/125309814/65176465-058c-4474-9401-fd70dcefdd07) |
| ![Screenshot 2024-02-16 at 12 24 05 PM](https://github.com/linode/manager/assets/125309814/4e7b1b1e-f94a-4b62-a4fa-01fb33492e9a) | ![Screenshot 2024-02-16 at 12 23 33 PM](https://github.com/linode/manager/assets/125309814/98a3eb1d-f9e3-4b3a-afbc-34aef4bb6bd4) |![Screenshot 2024-02-16 at 2 38 13 PM](https://github.com/linode/manager/assets/125309814/92b38bef-9414-4a34-b57a-bf4fd23a816a) |

## How to test 🧪

### Prerequisites
(How to setup test environment)
- Log into two accounts side by side:
   - An unrestricted admin user account: full access
   - A restricted user account (use Incognito for this)
       - Start with `Read Only` for everything

### Reproduction steps
(How to reproduce the issue, if applicable)
- Observe that before changes, a user can access and even begin filling out form information before eventually getting denied by the API. We want to avoid this.

### Verification steps
(How to verify changes)
- Test these flows as a restricted user with 'read only' access
- Then turn on MSW and Parent/Child feature flag
    - As a parent, observe that you cannot edit company name
- Set `user_type` to `child` in [serverHandlers](https://github.com/linode/manager/blob/develop/packages/manager/src/mocks/serverHandlers.ts#L554)  
    - Observe that you cannot access any of these drawers, and if you navigate via the URL's you get appropriate error notices with disabled fields

1. http://localhost:3000/account/billing/edit
2. http://localhost:3000/account/billing/add-payment-method 
3. http://localhost:3000/account/billing/make-payment

## As an Author I have considered 🤔

*Check all that apply*
